sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,347 @@
+                    identifierReference.Parent is AssignmentExpressionSyntax)
+                {
+                    paramData.ShouldReportOn = false;
+                    continue;
+                }
+
+                var symbolUsedAs = FindParameterUseAsType(identifierReference, context.SemanticModel);
+                if (symbolUsedAs != null)
+                {
+                    paramData.AddUsage(symbolUsedAs);
+                }
+            }
+
+            return parametersToCheck
+                .Values
+                .Select(p => p.GetRuleViolation())
+                .WhereNotNull();
+        }
+
+        private static bool IsTrackedParameter(IParameterSymbol parameterSymbol)
+        {
+            var type = parameterSymbol.Type;
+
+            return !type.DerivesFrom(KnownType.System_Array) &&
+                   !type.IsValueType &&
+                   !type.Is(KnownType.System_String);
+        }
+
+        private SyntaxNode GetNextUnparethesizedParent(SyntaxNode node)
+        {
+            var expression = node as ExpressionSyntax;
+            if (expression == null)
+            {
+                return node;
+            }
+
+            var topmostParent = SyntaxHelper.GetSelfOrTopParenthesizedExpression(expression);
+
+            return (topmostParent is ParenthesizedExpressionSyntax)
+                ? topmostParent.Parent
+                : topmostParent;
+        }
+
+        private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax idRef, SemanticModel semanticModel)
+        {
+            var idRefParent = GetNextUnparethesizedParent(idRef.Parent);
+
+            var conditionalAccess = idRefParent as ConditionalAccessExpressionSyntax;
+            if (conditionalAccess != null)
+            {
+                var binding = conditionalAccess.WhenNotNull as MemberBindingExpressionSyntax;
+
+                if (binding != null)
+                {
+                    var name = binding.Name;
+                    if (name == null)
+                    {
+                        return null;
+                    }
+
+                    var accessedMember = semanticModel.GetSymbolInfo(name).Symbol;
+                    return HandlePropertyOrField(idRef, accessedMember);
+                }
+
+                var invocationExpression = conditionalAccess.WhenNotNull as InvocationExpressionSyntax;
+                if (invocationExpression != null)
+                {
+                    var memberBinding = invocationExpression.Expression as MemberBindingExpressionSyntax;
+                    if (memberBinding != null)
+                    {
+                        var invocationSymbol = semanticModel.GetSymbolInfo(memberBinding).Symbol;
+                        return HandleInvocation(idRef, invocationSymbol, semanticModel);
+                    }
+                }
+            }
+            else if (idRefParent is MemberAccessExpressionSyntax)
```

Comment:
Previous style is `as` (on the if) and there you use `is`, I'd rather make sure we keep the same pattern everywhere inside a method.